crates/pairwise_nostd_internal/src/accumulator.rs

```diff
@@ -30,86 +32,187 @@
+/// I think that there are significant advantages to improving the readability
+/// of the code and it would be relatively move away from using the struct
+/// representation later (it would be easier to do that before we start
+/// introducing more accumulators).
+pub struct DataElement {
+    pub value: f64,
+    pub weight: f64,
+}
 
-/// describes the output components from a single Accumulator statepack
+/// describes the output components from a single Accumulator accum_state
 pub enum OutputDescr {
     MultiScalarComp(&'static [&'static str]),
     SingleVecComp { size: usize, name: &'static str },
 }
 
 impl OutputDescr {
     /// the number of components to allocate per component
-    pub fn n_per_statepack(&self) -> usize {
+    pub fn n_per_accum_state(&self) -> usize {
         match self {
             Self::MultiScalarComp(names) => names.len(),
             Self::SingleVecComp { size, .. } => *size,
         }
     }
 }
 
-// Accumulators generally operate on individual state-packs (the implementation
-// is currently inefficient, but we will refactor and try to come up with
-// better abstractions once we are done mapping out all the requirements)
-//
-// In the context of the larger package, there will be n accumulators
+/// Accumulators generally operate on individual `accum_state`s.
+///
+/// The implementation is currently inefficient, but we will refactor and try
+/// to come up with better abstractions once we are done mapping out all the
+/// requirements.
+///
+/// # Note
+/// It might be elegant to:
+/// - introduce a separate, related trait, called `Consume` or `UpdateLeft`,
+///   which would look something like
+///   ```text
+///   pub trait UpdateLeft<Rhs> {
+///       fn update_left(accum_state: &mut ArrayViewMut1<f64>, right: Rhs);
+///   }
+///   ```
+/// - rename the current trait to something like `AccumMiscOps` and delete the
+///   `consume` and `merge` methods.
+/// - define the `Accumulator` trait as a super trait:
+///   ```text
+///   pub trait Accumulator:
+///       AccumMiscOps
+///       + UpdateLeft<&Self>
+///       + UpdateLeft<&DataElement>
+///   {}
+///   ```
+/// - thus for each Accumulator, we would be implementing `UpdateLeft<&Self>`
+///   (instead of `merge`) and `UpdateLeft<&DataElement>` (instead of
+///   `consume`).
+/// - The advantage to doing this is that `update_left` would be used for
+///   merges and consuming -- it's purely aesthetic. It may also be nice if we
+///   introduce the idea of partially digested data... (it probably isn't worth
+///   the effort)
 pub trait Accumulator {
     /// the number of f64 elements needed to track the accumulator data
-    fn statepack_size(&self) -> usize;
+    fn accum_state_size(&self) -> usize;
 
     /// initializes the storage tracking the acumulator's state
-    fn reset_statepack(&self, statepack: &mut ArrayViewMut1<f64>);
+    fn reset_accum_state(&self, accum_state: &mut AccumStateViewMut);
```

Comment:
We might want to call this `initialize_accum_state` or `zero_accum_state` or similar, because as a user, I wouldn't think to `reset` after creating an object.  That said, those aren't ideal either and this can be addressed with documentation.

Reply:
I choose `init_accum_state`. It's unfortunate that you have to explicitly initialize it... But I think we have to do it to maintain flexibility over the data-layout.